common/beerocks/bwl/shared/netlink_socket.h

```diff
@@ -0,0 +1,112 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * Copyright (c) 2020 MaxLinear
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+#ifndef __BWL_NETLINK_SOCKET_H__
+#define __BWL_NETLINK_SOCKET_H__
+
+#include <functional>
+#include <memory>
+
+/**
+ * Forward declarations
+ */
+struct nl_sock;
+struct nl_msg;
+
+namespace bwl {
+
+/**
+ * @brief C++ wrapper class around the core library of the Netlink Protocol Library Suite (libnl).
+ *
+ * The Netlink socket is Linux kernel interface used for inter-process communication (IPC) between
+ * both the kernel and userspace processes.
+ *
+ * The Netlink core library contains C functions implementing the fundamentals required to use the
+ * Netlink protocol such as socket handling, message construction and parsing, and sending and
+ * receiving of data.
+ *
+ * This is a very simple C++ wrapper class around the Netlink core library. Its aim is to
+ * facilitate the use of the library from C++ applications. The class can be used as is or extended
+ * by derived classes to support extended versions of the netlink protocol like, for example, the
+ * generic netlink protocol.
+ *
+ * Known derived classes: nl_genl_socket.
+ */
+class netlink_socket {
+
+public:
+    /**
+     * @brief Class constructor.
+     *
+     * Allocates the Netlink socket structure.
+     *
+     * @param protocol Netlink protocol to use. Possible values are listed in <linux/netlink.h>
+     */
+    explicit netlink_socket(int protocol);
```

Comment:
If you try to create a `netlink_socket` using default constructor, then you get a compilation error because it's not defined:
```
error: no matching function for call to 'bwl::netlink_socket::netlink_socket()'
```
If you try to create a `netlink_socket`using copy constructor, then you also get a compilation error because such constructor is implicitly deleted due to the unique_ptr member.
```
error: use of deleted function 'bwl::netlink_socket::netlink_socket(const bwl::netlink_socket&)'
```
I can however add those constructors if there is any coding style rule about this.